ethers-solc/src/resolver.rs

```diff
@@ -0,0 +1,518 @@
+//! Resolution of the entire dependency graph for a project.
+//!
+//! This module implements the core logic in taking all contracts of a project and creating a
+//! resolved graph with applied remappings for all source contracts.
+//!
+//! Some constraints we're working with when resolving contracts
+//!
+//!   1. Each file can contain several source units and can have any number of imports/dependencies
+//! (using the term interchangeably). Each dependency can declare a version range that it is
+//! compatible with, solidity version pragma.
+//!   2. A dependency can be imported from any directory,
+//! see `Remappings`
+//!
+//! Finding all dependencies is fairly simple, we're simply doing a DFS, starting the source
+//! contracts
+//!
+//! ## Performance
+//!
+//! Note that this is a relatively performance-critical portion of the ethers-solc preprocessing.
+//! The data that needs to be processed is proportional to the size of the dependency
+//! graph, which can, depending on the project, often be quite large.
+//!
+//! Note that, unlike the solidity compiler, we work with the filesystem, where we have to resolve
+//! remappings and follow relative paths. We're also limiting the nodes in the graph to solidity
+//! files, since we're only interested in their
+//! [version pragma](https://docs.soliditylang.org/en/develop/layout-of-source-files.html#version-pragma),
+//! which is defined on a per source file basis.
+
+use std::{
+    collections::{HashMap, VecDeque},
+    path::{Component, Path, PathBuf},
+};
+
+use rayon::prelude::*;
+use semver::VersionReq;
+use solang::parser::pt::{Import, SourceUnitPart};
+
+use crate::{error::Result, utils, ProjectPathsConfig, Solc, Source, Sources};
+
+/// Represents a fully-resolved solidity dependency graph. Each node in the graph
+/// is a file and edges represent dependencies between them.
+/// See also https://docs.soliditylang.org/en/latest/layout-of-source-files.html?highlight=import#importing-other-source-files
+#[derive(Debug)]
+pub struct Graph {
+    nodes: Vec<Node>,
+    /// The indices of `edges` correspond to the `nodes`. That is, `edges[0]`
+    /// is the set of outgoing edges for `nodes[0]`.
+    edges: Vec<Vec<usize>>,
+    /// index maps for a solidity file to an index, for fast lookup.
+    indices: HashMap<PathBuf, usize>,
+    /// with how many input files we started with, corresponds to `let input_files =
+    /// nodes[..num_input_files]`.
+    num_input_files: usize,
+    /// the root of the project this graph represents
+    #[allow(unused)]
+    root: PathBuf,
+}
+
+impl Graph {
+    /// Returns a list of nodes the given node index points to for the given kind.
+    pub fn imported_nodes(&self, from: usize) -> &[usize] {
+        &self.edges[from]
+    }
```

Comment:
what do we get from exposing this as a method?

Reply:
can be used to iterate over all imports of a file.
this will be useful when I add tree printing.
but can be private for now